framework/src/araui/libs/upng.c

```diff
@@ -0,0 +1,1299 @@
+	}
+
+	/* scan through the chunks again, this time copying the values into
+	 * our compressed buffer.  there's no reason to validate anything a second time. */
+	chunk = upng->source.buffer + 33;
+	while (chunk < upng->source.buffer + upng->source.size) {
+		unsigned long length;
+		const unsigned char *data;	/*the data in the chunk */
+
+		length = upng_chunk_length(chunk);
+		data = chunk + 8;
+
+		/* parse chunks */
+		if (upng_chunk_type(chunk) == CHUNK_IDAT) {
+			memcpy(compressed + compressed_index, data, length);
+			compressed_index += length;
+		} else if (upng_chunk_type(chunk) == CHUNK_IEND) {
+			break;
+		}
+
+		chunk += upng_chunk_length(chunk) + 12;
+	}
+
+	/* allocate space to store inflated (but still filtered) data */
+	inflated_size = ((upng->width * (upng->height * upng_get_bpp(upng) + 7)) / 8) + upng->height;
+	inflated = (unsigned char*)malloc(inflated_size);
+	if (inflated == NULL) {
+		free(compressed);
+		SET_ERROR(upng, UPNG_ENOMEM);
+		return upng->error;
+	}
+
+	/* decompress image data */
+	error = uz_inflate(upng, inflated, inflated_size, compressed, compressed_size);
+	if (error != UPNG_EOK) {
+		free(compressed);
+		free(inflated);
+		return upng->error;
+	}
+
+	/* free the compressed compressed data */
+	free(compressed);
+
+	/* allocate final image buffer */
+	upng->size = (upng->height * upng->width * upng_get_bpp(upng) + 7) / 8;
+	upng->buffer = (unsigned char*)malloc(upng->size);
+	if (upng->buffer == NULL) {
+		free(inflated);
+		upng->size = 0;
+		SET_ERROR(upng, UPNG_ENOMEM);
+		return upng->error;
+	}
+
+	/* unfilter scanlines */
+	post_process_scanlines(upng, upng->buffer, inflated, upng);
+	free(inflated);
+
+	if (upng->error != UPNG_EOK) {
+		free(upng->buffer);
+		upng->buffer = NULL;
+		upng->size = 0;
+	} else {
+		upng->state = UPNG_DECODED;
+	}
+
+	/* we are done with our input buffer; free it if we own it */
+	upng_free_source(upng);
+
+	return upng->error;
+}
+
+static upng_t* upng_new(void)
+{
+	upng_t* upng;
+
+	upng = (upng_t*)malloc(sizeof(upng_t));
+	if (upng == NULL) {
+		return NULL;
+	}
+
+	upng->buffer = NULL;
+	upng->size = 0;
+
+	upng->width = upng->height = 0;
+
+	upng->color_type = UPNG_RGBA;
+	upng->color_depth = 8;
+	upng->format = UPNG_RGBA8;
+
+	upng->state = UPNG_NEW;
+
+	upng->error = UPNG_EOK;
+	upng->error_line = 0;
+
+	upng->source.buffer = NULL;
+	upng->source.size = 0;
+	upng->source.owning = 0;
+
+	return upng;
+}
+
+upng_t* upng_new_from_bytes(const unsigned char* buffer, unsigned long size)
+{
+	upng_t* upng = upng_new();
+	if (upng == NULL) {
+		return NULL;
+	}
+
+	upng->source.buffer = buffer;
+	upng->source.size = size;
+	upng->source.owning = 0;
+
+	return upng;
+}
+
+upng_t* upng_new_from_file(const char *filename)
+{
+	upng_t* upng;
+	unsigned char *buffer;
+	FILE *file;
+	long size;
+
+	upng = upng_new();
+	if (upng == NULL) {
+		return NULL;
+	}
+
+	file = fopen(filename, "rb");
+	if (file == NULL) {
+		SET_ERROR(upng, UPNG_ENOTFOUND);
+		return upng;
```

Comment:
In my opinion, You have to free upng and return NULL instead of dummy data of upng..
Also errno will help you to get reason why open failed.